cmd/azure.go

```diff
@@ -0,0 +1,363 @@
+func (a AzureObjects) NewMultipartUpload(bucket, object string, metadata map[string]string) (uploadID string, err error) {
+	// Azure doesn't return a unique upload ID and we use object name in place of it. Azure allows multiple uploads to
+	// co-exist as long as the user keeps the blocks uploaded (in block blobs) unique amongst concurrent upload attempts.
+	// Each concurrent client, keeps its own blockID list which it can commit.
+	uploadID = object
+	return uploadID, azureToObjectError(traceError(a.client.CreateBlockBlob(bucket, object)))
+}
+
+// CopyObjectPart - Not implemented.
+func (a AzureObjects) CopyObjectPart(srcBucket, srcObject, destBucket, destObject string, uploadID string, partID int, startOffset int64, length int64) (info PartInfo, err error) {
+	return info, traceError(NotImplemented{})
+}
+
+// Encode partID+md5Hex to a blockID.
+func azureGetBlockID(partID int, md5Hex string) string {
+	return base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%.5d.%s", partID, md5Hex)))
+}
+
+// Decode blockID to partID+md5Hex.
+func azureParseBlockID(blockID string) (int, string, error) {
+	idByte, err := base64.StdEncoding.DecodeString(blockID)
+	if err != nil {
+		return 0, "", traceError(err)
+	}
+	idStr := string(idByte)
+	splitRes := strings.Split(idStr, ".")
+	if len(splitRes) != 2 {
+		return 0, "", traceError(errUnexpected)
+	}
+	partID, err := strconv.Atoi(splitRes[0])
+	if err != nil {
+		return 0, "", traceError(err)
+	}
+	return partID, splitRes[1], nil
+}
+
+// PutObjectPart - Use Azure equivalent PutBlockWithLength.
+func (a AzureObjects) PutObjectPart(bucket, object, uploadID string, partID int, size int64, data io.Reader, md5Hex string, sha256sum string) (info PartInfo, err error) {
+	id := azureGetBlockID(partID, md5Hex)
+	err = a.client.PutBlockWithLength(bucket, object, id, uint64(size), data, nil)
+	if err != nil {
+		return info, azureToObjectError(traceError(err), bucket, object)
+	}
+	info.PartNumber = partID
+	info.ETag = md5Hex
+	info.LastModified = time.Now().UTC()
+	info.Size = size
+	return info, nil
+}
+
+// ListObjectParts - Use Azure equivalent GetBlockList.
+func (a AzureObjects) ListObjectParts(bucket, object, uploadID string, partNumberMarker int, maxParts int) (result ListPartsInfo, err error) {
+	resp, err := a.client.GetBlockList(bucket, object, storage.BlockListTypeUncommitted)
+	if err != nil {
+		return result, azureToObjectError(traceError(err), bucket, object)
+	}
+	for _, part := range resp.UncommittedBlocks {
+		partID, md5Hex, err := azureParseBlockID(part.Name)
+		if err != nil {
+			return result, err
+		}
+		result.Parts = append(result.Parts, PartInfo{
+			partID,
+			time.Now().UTC(),
+			md5Hex,
+			part.Size,
+		})
+	}
+	result.Bucket = bucket
+	result.Object = object
+	result.UploadID = uploadID
+	result.MaxParts = maxParts
+	return result, nil
+}
+
+// AbortMultipartUpload - Not Implemented.
+func (a AzureObjects) AbortMultipartUpload(bucket, object, uploadID string) error {
```

Comment:
Uncommitted blocks in a blob are automatically deleted after a week. However, for reference, in case immediate removal of uncommitted blocks is desired (as they are charged for a week), one way I could find was using copy-object with source name = destination name:

> You can copy a source blob to a destination blob with the same name, effectively replacing the destination blob. Such a copy operation removes any uncommitted blocks and overwrites the blob's metadata.

from [here](https://docs.microsoft.com/en-us/rest/api/storageservices/fileservices/copy-blob#remarks).